memdown.js

```diff
@@ -116,6 +117,29 @@ MemIterator.prototype._test = function () {
   return true
 }
 
+MemIterator.prototype._seek = function (target) {
+  var options = this._options
+
+  var tree = this.db._store
+
+  target = target.toString()
+
+  if (this._options.asBuffer === false) {
+    target = Buffer.from(target)
+  }
+
+  if (ltgt.filter(options)(target) === false) {
```

Comment:
Can we use the existing `this._test` functions? Because `ltgt.filter()` won't work with custom comparators (which I'm about to introduce) and internally uses `undefined` as not-defined signal (which works but I still prefer using [our own explicit signal](https://github.com/Level/memdown/blob/cd26ada6fbd391d9c6d485fedd35ca0535d4c444/memdown.js#L11)).

Reply:
might need to rework that then; as you could potentially seek backwards? so you'd need to know the lower and upper bounds?

Reply:
Maybe we can trick it by flipping some condition, something like `if (this._test(target) ? !reverse : reverse)`.

Reply:
I don't think so, as you could seek backwards beyond the (other-side) limit and there would be no way to know it was invalid.

Reply:
> as you could potentially seek backwards(?)

Seeking backwards (or precisely seeking reverse the iterating direction) is perfectly fine in `leveldb`, and is *unintentionally* supported in `leveldown`.

As far as I can remember, we didn't promise this behavior in any document. Before [introducing .seek() to abstract-leveldown](https://github.com/Level/abstract-leveldown/commit/d806bf147d12186990db7fdc6da4c95e7a558ccc#diff-756c4775324d3d60cf94b32a133f5fb4) it was impossible to seek backwards using the stream interface with `levelup`. Maybe it's time for us to make this clear?

Reply:
@peakji I assumed that seeking backwards just worked (tm) in `leveldown` and I don't think there's a need to document that, but adding an abstract test would be good.

Reply:
@MeirionHughes How about this: make a copy of `iterator._tree` (which already has one bound applied to it) in the constructor. Use that tree as a starting point instead of `db._store`. Then use `_test` for the other bound.

Reply:
@MeirionHughes oof, this works:

```js
MemIterator.prototype._outOfRange = function (target) {
  if (!this._test(target)) {
    return true
  } else if (this._lowerBound === NONE) {
    return false
  } else if (!this._reverse) {
    if (ltgt.lowerBoundInclusive(this._options)) {
      return ltgt.compare(target, this._lowerBound) < 0
    } else {
      return ltgt.compare(target, this._lowerBound) <= 0
    }
  } else {
    if (ltgt.upperBoundInclusive(this._options)) {
      return ltgt.compare(target, this._lowerBound) > 0
    } else {
      return ltgt.compare(target, this._lowerBound) >= 0
    }
  }
}

MemIterator.prototype._seek = function (target) {
  // TODO
  target = target.toString()

  if (this._outOfRange(target)) {
    this._tree = this.db._store.end
    this._tree.next()
  } else if (this._reverse) {
    this._tree = this.db._store.le(target)
  } else {
    this._tree = this.db._store.ge(target)
  }
}
```

Reply:
updated - do I need / is this okay?

```
  if (this._options.asBuffer !== true) {
    target = String(target)
  }
```

Reply:
`*asBuffer` options should not be applied to input (https://github.com/Level/memdown/pull/184#issuecomment-504316006), but to make tests pass, we do need `target = String(target)` for now

Reply:
Alternatively, we could skip the seek tests for now and don't convert the target in any way.

Reply:
So there is nothing in memdown to dictate whether its storing Buffer or string? it could just be either? 

Yeah - I guess we'd have to disable the seek tests for now as `target = String(target)` could screw up if you're storing/comparing Buffers.

Reply:
> So there is nothing in memdown to dictate whether its storing Buffer or string? it could just be either?

Yes. The user is responsible for what goes in; `memdown` does not convert any inputs.

> I guess we'd have to disable the seek tests for now 

OK, let's do that.

Reply:
I've tried to keep the tests, but avoid polluting the runtime with: 
```
// temp fix for seek tests
var baseSeek = MemIterator.prototype._seek
MemIterator.prototype._seek = function (target) {
  return baseSeek.call(this, String(target))
}
```

Reply:
Ah, that's clever.